core/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    void checkForRemovedNodes(ClusterChangedEvent event) {
+        event.nodesDelta().removedNodes().stream().forEach(node -> {
+            try (Releasable ignored = nodeLocks.acquire(node)) {
+                Integer current = nodes.remove(node);
+                assert current != null : "node " + node + " was removed in event but not in internal nodes";
+                transportService.disconnectFromNode(node);
+            } catch (Throwable e) {
+                logger.warn("failed to disconnect to node [" + node + "]", e);
+            }
+        });
+    }
+
+    void validateNodeConnected(DiscoveryNode node) {
+        assert nodeLocks.isHeldByCurrentThread(node) : "validateNodeConnected must be called under lock";
+        if (lifecycle.stoppedOrClosed()) {
+            return;
+        }
+        if (localNode.shouldConnectTo(node) == false) {
+            return;
+        }
+        try {
+            // we double check existence of node since connectToNode might take time...
+            if (nodes.containsKey(node) == false) {
+                return;
+            }
+
+            if (transportService.nodeConnected(node)) {
+                nodes.put(node, 0);
+                return;
+            }
+            transportService.connectToNode(node);
+            nodes.put(node, 0);
+        } catch (Exception e) {
+            Integer nodeFailureCount = nodes.get(node);
+            assert nodeFailureCount != null : node + " didn't have a counter in nodes map";
+            nodeFailureCount = nodeFailureCount + 1;
+            // log every 6th failure
+            if ((nodeFailureCount % 6) == 1) {
+                logger.warn("failed to connect to node {} (tried [{}] times)", e, node, nodeFailureCount);
+            }
+            nodes.put(node, nodeFailureCount);
+        }
+    }
+
+    class ConnectionChecker extends AbstractRunnable {
+
+        @Override
+        public void onFailure(Throwable t) {
+            logger.warn("unexpected error while checking for node reconnects", t);
+        }
+
+        protected void doRun() {
+            for (DiscoveryNode node : nodes.keySet()) {
+                try (Releasable ignored = nodeLocks.acquire(node)) {
+                    validateNodeConnected(node);
+                }
+            }
+        }
+
+        @Override
+        public void onAfter() {
+            if (lifecycle.started()) {
+                backgroundFuture = threadPool.schedule(reconnectInterval, ThreadPool.Names.GENERIC, this);
+            }
+        }
+    }
+
+    @Override
+    protected void doStart() {
+        clusterService.addFirst(this::checkForAddedNodes);
+        clusterService.addLast(this::checkForRemovedNodes);
+        backgroundFuture = threadPool.schedule(reconnectInterval, ThreadPool.Names.GENERIC, new ConnectionChecker());
+    }
+
+    @Override
+    protected void doStop() {
+        FutureUtils.cancel(backgroundFuture);
+        clusterService.remove(this::checkForAddedNodes);
```

Comment:
Tja. It feels like the "clean thing to do". Agree it doesn't have much impact. Other places do this as well....